WikiWatershed/model-my-watershed
# Asynchronize Analyze

## Overview

Given that the **Analyze** tab is getting more content, we switch to an asynchronous fetching pattern so that quicker results may return sooner than slower ones.

The central difference is that instead of one analyze endpoint which returns all results, we now have multiple analyze endpoints which return multiple results. This manifests as follows in our code:

 * The main `/api/modeling/start/analyze/` endpoint now returns only the results of the original Analyze chain, which includes counts from NLCD and Soil rasters. The results of `animal_population` and `point_source_pollution` have been removed from its output.
 * We now have two new endpoints `/api/modeling/start/analyze/animals/` and `api/modeling/start/analyze/pointsources/` for Animals and Point Sources respectively. Both operate identically to the `analyze` endpoint, inasmuch they respond with a job id which can be polled to get results. Since both are quick jobs, the results are usually available on the first poll.
 * Instead of having one `App.analyzeModel` which was the task runner for the `analyze` endpoint, we now have `App.analyzeCollection` which contains task runners for `analyze`, `analyze/animals`, and `analyze/pointsources`.
 * Instead of a number of `if`s and `switch`es in `TabContentView`, we now show a different result view for each result type. This separation allows us to more easily make changes between result views.
 * Instead of the entire `AnalyzeWindow` handling the polling messaging, that is now handled by `TabContentView` so we can show different polling states for each task runner. The polling is also reflected in a spinner on the tab panel.
 * Each result view is populated by a view model, which points to exactly one task runner. The `land` and `soil` view models point to the same task runner, since they get their results from the same endpoint, and thus are updated at the same time. `animals` and `pointsource` view models point to their own task runners, and are updated independently. These can finish or fail independently without affecting one another.

### Demo

http://gfycat.com/BasicElaborateGopher

### Notes

~~The tests for the Analyze view were never very elaborate. Now they are testing views that are no longer used. I tried to come up with better ones, but that's a big task on its own. We should probably create a card for that.~~

The new tests exercise only the success workflow. We should add tests that exercise waiting, failing workflows as well.

## Testing Instructions

 * Bundle scripts `./scripts/bundle.sh --debug`
 * Debug Celery so it picks up new tasks `./scripts/debugcelery.sh`
 * Open the app. Draw / Select an area of interest.
 * In the Analyze sidebar, instead of a simple "Analyzing" message, you should see four tabs corresponding to each result type, each with a spinner and its own "Analyzing" message. The Animals and Point Source tabs should populate first, then Land and Soil should populate together.
 * Going to Model stage should preserve the Analyze tab and it should continue to work as before.
 * Going to Draw and going back to Analyze without changing the area of interest should not trigger new requests.
 * Loading a saved project should trigger Analyze requests, and they should behave identically as they did when coming in from Draw.

Connects #1438
